### DISCUSSION TITLE
Headless tulip

### DISCUSSION THREAD

**Discussion by smutax** [shorepine/tulipcc#115] [General]
Hi. Is there any headless (w/o screen) use case developers have in mind for tulipcc? Thanks
Regards
ViBe

**Comment by bwhitman** [comment]
It depends what you're hoping to get out of Tulip! If you want the sound capabilities, I'd look at [Alles](https://github.com/bwhitman/alles). If you want just a Python prompt, the bare Micropython esp32 project would help you. Tulip is basically Alles + a display core + MIDI + Micropython.

**Reply by smutax** [reply]
For myself,  i see Alles/AMY/Tulip as this fantastic midi-controlable synth module! Sorry if i oversimplify your vision behind the project.

**Reply by bwhitman** [reply]
Not at all! Tulip is the only one with direct MIDI support at the moment. If you want to use Tulip without a screen, you can just ensure `tulip.tfb_log_start()` is on, and you can type into python via UART repl. Or, you can back-port MIDI onto Alles and control it that way.

**Reply by smutax** [reply]
That's great! Thanks Brian. Talking of midi, what is the scope and current state of Tulip's midi implementation?

**Reply by dpwe** [reply]
> Talking of midi, what is the scope and current state of Tulip's midi implementation?

On the Tulip board, MIDI in/out works over the two minijack connectors (the third is stereo audio out).  You can write raw MIDI messages to the output with `tulip.midi_out(bytes)`, and you can set a Python function to be called whenever there is new data available from `tulip.midi_in()`.  See the [MIDI section in the API documentation](https://github.com/dpwe/tulipcc/blob/main/docs/tulip_api.md#midi).

MIDI input is also possible by connecting a USB-MIDI keyboard to the USB-A "keyboard connector" on the Tulip board.  We now have hot-swapping working, so you can unplug your QWERTY keyboard and plug in your MIDI keyboard.  However, you cannot use a USB Hub to connect them both at the same time - the USB hosting stack doesn't support hubs.

Tulip Desktop on the Mac "just works" with USB-MIDI devices in my experience.  As I understand, Tulip Desktop on Linux has not yet been connected to a Linux MIDI driver.

**Reply by smutax** [reply]
Thanks Dan. Perhaps i wasn't clear saying "midi implementation". What i mean, will it be possible to control and play Tulip(Amy) using standard midi messages, like Program change, Control change, Note on/off, Sysex etc.? I guess you have answered already, pointing to Python function upon midi in activity. Cheers.

**Reply by dpwe** [reply]
Yes, it's something you can just do in Python. I've been working on code to allow a MIDI keyboard to play the Tulip like a DX7 (including program changes): [polysynth.py](https://github.com/dpwe/tulip_programs/blob/main/polysynth.py)

**Reply by smutax** [reply]
Thanks. That's very interesting! Enjoyed the video demo of Xanadu as well!

**Reply by RocketManRC** [reply]
I know this is an old thread but I am new to Tulip and very interested in this headless topic. With the hints provided here I have been able to get an initial headless system working with an ESP32S3 N16R8 dev board and a PCM5102A DAC from AliExpress. Access to the MicroPython REPL is serial via the USB-C COM connector and USB MIDI works via the USB-C USB port. Since the devil of these things is always in the details I'm happy to write it up if others are still interested.

![HeadlessTulip](https://github.com/user-attachments/assets/93fa4d06-e469-40aa-aaea-5732dfd66807)

**Reply by smutax** [reply]
Hi Rick. This is great! I'm very much interested still! Am looking forward to your write up and would like to dig into those details. Feel free to include as much information as you like including links to AliExpress purchases that worked for you. Many thanks.

**Reply by bwhitman** [reply]
Very cool! I hadn't really thought about this use case, but I bet it's a great way to make headless synths in Python that work with MIDI.

**Reply by bwhitman** [reply]
I do wonder if we should have an optional mode that uses the native Micropython UART REPL instead of our ... modified REPL. I assume you have to call `tulip.tfb_log_start()` to see the REPL in the UART?

**Reply by RocketManRC** [reply]
Yes I put tulip.tfb_log_start() in boot.py and that enables the REPL. The rest is just the details about configuring pins and setting up the DAC properly. I wasn't really expecting that USB MIDI would work out of the box but it does which is great.

My original interest in Tulip was to maybe add sound generation to some touch based MIDI keyboards that I designed but now I'm realizing that many more things are possible.

What a great project!

**Reply by bwhitman** [reply]
I'd also suggest tulip.display_stop() if running headless. You'll have more headroom for polyphony!

**Reply by RocketManRC** [reply]
@smutax I have put my initial attempt at documenting this on my Github and would appreciate if you have a look at it and let me know if the level of detail is what you were looking for:

https://github.com/RocketManRC/tulipcc-headless?tab=readme-ov-file

I haven't put links to Aliexpress but will do that.

**Reply by bwhitman** [reply]
This is great! Re:

> The only changes required are to add 2 lines to boot.py but it means that you have to build the firmware yourself.

That's not true, you can edit `boot.py` once you've loaded Tulip from a precompiled binary. Just start the binary, and type (blindly) `tulip.tfb_log_start()`, then enter, then you'll start seeing the Python prompt. Then do 

```python
>>> bootpy = "import tulip\ntulip.tfb_log_start()\ntulip.display_stop()\n"
>>> b = open("boot.py","w")
>>> b.write(bootpy)
>>> b.close()
```

**Reply by RocketManRC** [reply]
Thanks Brian that works perfectly and eliminates a lot of overhead to get started!

I will update my document.

**Reply by RocketManRC** [reply]
I've updated the document to include flashing the firmware with a precompiled binary which is by far less effort and works just fine. I've also added a couple of links to components on AliExpress.

The link to the document is the same but here it is again:

https://github.com/RocketManRC/tulipcc-headless?tab=readme-ov-file

I hope this is helpful and any questions or suggestions are welcome!

**Reply by smutax** [reply]
> I've updated the document to include flashing the firmware with a precompiled binary which is by far less effort and works just fine. I've also added a couple of links to components on AliExpress.
> 
> The link to the document is the same but here it is again:
> 
> https://github.com/RocketManRC/tulipcc-headless?tab=readme-ov-file
> 
> I hope this is helpful and any questions or suggestions are welcome!

Nicely done Rick! Thanks a lot for that, I appreciate! As always big up the developers for their amazing project(s). Cheers.
